sage.py

```diff
@@ -571,393 +485,271 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+        for (attacker, victim), episodes in team.items():
             if len(episodes) < 2:
                 continue
-            perp= attacker[0]
-            vic = attacker[1]
-            #print(perp)
-            #if ('10.0.0' in perp or '10.0.1' in perp):
+            # if ('10.0.0' in attacker or '10.0.1' in attacker):
             #        continue
             
-            att = 't'+str(tid)+'-'+perp
-            #print(att)
+            att = 't' + str(tid) + '-' + attacker
             if att not in host_data.keys():
-                host_data[att] = []  
-            #EPISODE DEF: (startTime, endTime, mcat, len(rawevents), volume(alerts), epiPeriod, epiServices, list of unique signatures)
-            ext = [(x[0], x[1], x[2], x[3], x[4], x[5], x[6], x[7], x[8], vic) for x in episodes]
+                host_data[att] = []
 
-            host_data[att].append(ext)
+            extended_episode = [epi + (victim,) for epi in episodes]
+
+            host_data[att].append(extended_episode)
             host_data[att].sort(key=lambda tup: tup[0][0])
             
     print('\n# episode sequences ', len(host_data))   
+    return host_data
+
 
-    team_strat = list(host_data.values())
-    #print(len(team_strat[0]), [(len(x)) for x in team_strat[0]])
-    #print([[a[2] for a in x] for x in team_strat[0]])  
-    return (host_data)        
-
-def break_into_subbehaviors(host_data):        
-    attackers = []
-    keys = []
-    alerts = []
-    cutlen = 4
-    FULL_SEQ= False
+# Split episode sequences for an attacker-victim pair into episode subsequences.
+# Each episode subsequence represents an attack attempt.
+def break_into_subbehaviors(host_data):
+    attacks = []
+    subsequences = []
+    cut_length = 4
+    FULL_SEQ = False
         
-    #print(len(team))
     print('----- Sub-sequences -----')
-    for tid, (atta,victims) in enumerate(host_data.items()):
-        print((tid+1), sep=' ', end=' ', flush=True)
-        #print(atta)
-        #print(len(victims))
-        for episodes in victims:
+    for i, (attacker, victim_episodes) in enumerate(host_data.items()):
+        print((i + 1), sep=' ', end=' ', flush=True)
+        for episodes in victim_episodes:
             if len(episodes) < 2:
                 continue
+
             victim = episodes[0][-1]
-            pieces = math.floor(len(episodes)/cutlen)
-            _episodes = []
+            attack = attacker + '->' + victim
+            pieces = math.floor(len(episodes) / cut_length)
             if FULL_SEQ:
-                att = atta+'->'+victim
-                #print(att, [x[2] for x in episodes])
-                keys.append(att)
-                alerts.append(episodes)
+                attacks.append(attack)
+                subsequences.append(episodes)
+                continue
+            if pieces < 1:
+                attacks.append(attack + '-0')
+                subsequences.append(episodes)
+                continue
 
-            else:
```

Comment:
Large block in the diff, but essentially I have simplified some nested if-elses

Reply:
I expect you double checked that there are no missing edge cases?

Reply:
There should not be any edge cases, since instead of nesting blocks in the "else" part (if `FULL_SEQ` is false or if `pieces < 1` is false), I have added `continue`
![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/77510f02-f211-4b20-99ca-88030909c570)
![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/88478d72-dd32-4b04-b698-bbdaee03d0ba)

Reply:
As for the `if len(cuts) == 0:` part, the for loop `for j in range(len(cuts)):` will not execute, `subsequence` will be the same as `episodes` and `episodes < 2` have already been discarded above

![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/5cf7bb75-cb07-480d-a463-b4497ff2654a)
![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/5d8a1251-5868-4dc4-821f-00089631c730)

Reply:
As a check: I wrote subsequences to a file both before and after using CPTC-2017. The files are identical